scipy/interpolate/_bspl.pyx

```diff
@@ -457,9 +459,13 @@ def _make_design_matrix(const double[::1] x,
         ind = find_interval(t, k, xval, ind, extrapolate)
         _deBoor_D(&t[0], xval, k, ind, 0, &work[0])
 
-        data[(k + 1) * i : (k + 1) * (i + 1)] = work[:k + 1]
-        row_ind[(k + 1) * i : (k + 1) * (i + 1)] = i
-        col_ind[(k + 1) * i : (k + 1) * (i + 1)] = np.arange(
-            ind - k, ind + 1, dtype=int)
+        # data[(k + 1) * i : (k + 1) * (i + 1)] = work[:k + 1]
+        # indices[(k + 1) * i : (k + 1) * (i + 1)] = np.arange(ind - k, ind + 1)
+        for j in range((k + 1)):
+            data[(k + 1) * i + j] = work[j]
+            indices[(k + 1) * i + j] = ind - k + j
```

Comment:
Maybe let's see if using a memcpy for the `data[...] = work[:k+1]` is vilible in the profile. I suspect not but it'd be best to check while we're on the perf topic and benchmarking all this is on your fingertips.

Also can store `(k+1)*i  + j` in a variable and avoid recomputing it. Unlikely to help performance, but might be of a marginal readability improvement, to lump with other changes.

Reply:
This makes indeed a huge difference: 107 ms instead of ~10ms. I suspect that the cause is the many intermediate array allocations of `np.arange(ind - k, ind + 1)`. Also, it mixes Python and C(ython) code/calls.

Reply:
What I meant was to try memcpy from libc, which operates on raw pointers.

Reply:
Do you mean `memcopy(data, work, ..)`?  But how to fill `indices`? Unless I misunderstand something, I would still keep the current explicit for-loop over j.

Reply:
Yes, the loop to fill indices stays. Moving data and work arrays out of the loop *might* help at the level cache thrashing.

Reply:
```cython
memcpy(&data[(k + 1) * i], &work[0], (k + 1) * sizeof(double))
for j in range(k + 1):
    m = (k + 1) * i + j
    indices[m] = ind - k + j
```
gives
```python
%timeit -r 20 -n 100 BSpline.design_matrix(x, t, k)
13.1 ms ± 196 µs per loop (mean ± std. dev. of 20 runs, 100 loops each)
```
while
```cython
for j in range(k + 1):
    m = (k + 1) * i + j
    data[m] = work[j]
    indices[m] = ind - k + j
```
gives
```
11.8 ms ± 276 µs per loop (mean ± std. dev. of 20 runs, 100 loops each)
```
Binary size is the same.

Reply:
Ok, thanks! Good to know